File: Java/src/main/java/org/codecop/dependencies/d/RestCountriesAPI.java
package org.codecop.dependencies.d;

import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.mashape.unirest.http.HttpResponse;
import com.mashape.unirest.http.JsonNode;
import com.mashape.unirest.http.Unirest;
import org.codecop.dependencies.d.restcountries.CountryDescription;
import org.codecop.dependencies.d.restcountries.Root;

import java.util.ArrayList;
import java.util.List;
import java.util.Optional;

public class RestCountriesAPI {

    private static final Country HOME_BASE = new Country("AT");
    private static final String COUNTRY_INFORMATION_SERVICE_URL = "https://api.restcountries.com/countries/v5?pretty=1&limit=100";

    private static final RestCountriesAPI instance = new RestCountriesAPI();

    private final ObjectMapper objectMapper;

    private RestCountriesAPI() {
        this.objectMapper = new ObjectMapper();
    }

    public static RestCountriesAPI getInstance() {
        return instance;
    }

    public boolean isInCommonMarket(Country country) {
        final Optional<CountryDescription> countryDescription = getCountryDescriptionViaRestCall(country);
        return countryDescription.map(description -> description.getRegion().equals("Europe")).orElse(false);
    }

    public boolean isInAmericas(Country country) {
        final Optional<CountryDescription> countryDescription = getCountryDescriptionViaRestCall(country);
        return countryDescription.map(description -> description.getRegion().equals("Americas")).orElse(false);
    }

    public int distanceTo(Country country) {
        final Optional<CountryDescription> austria = getCountryDescriptionViaRestCall(HOME_BASE);
        final Optional<CountryDescription> other = getCountryDescriptionViaRestCall(country);

        if (!austria.isPresent() || !other.isPresent()) {
            throw new RestCountriesAPIException("Could not find country " + HOME_BASE + " or " + country);
        }

        return distBetween(austria.get().getCoordinates(), other.get().getCoordinates());
    }

    private int distBetween(CountryDescription.Coordinates fromCoordinates, CountryDescription.Coordinates toCoordinates) {
        return distBetween(fromCoordinates.lat, fromCoordinates.lng, toCoordinates.lat, toCoordinates.lng);
    }

    private int distBetween(double fromLatitude, double fromLongitude, double toLatitude, double toLongitude) {
        double earthRadius = 6371000; // meters
        double fLat = Math.toRadians(fromLatitude);
        double toLat = Math.toRadians(toLatitude);
        double diffLat = Math.toRadians(toLatitude - fromLatitude);
        double diffLng = Math.toRadians(toLongitude - fromLongitude);
        double a = Math.sin(diffLat / 2) * Math.sin(diffLat / 2) + Math.cos(fLat) * Math.cos(toLat) * Math.sin(diffLng / 2) * Math.sin(diffLng / 2);
        double c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));
        float dist = (float) (earthRadius * c);
        return (int) dist;
    }

    private Optional<CountryDescription> getCountryDescriptionViaRestCall(Country country) {
        return slowHttpCall().stream().filter(c -> c.codes.alpha_2.equals(country.toString())).findFirst();
    }

    public List<CountryDescription> slowHttpCall() {
        final List<CountryDescription> countryDescriptions = new ArrayList<>();

        try {
            for (int offset = 0; offset < 300; offset += 100) {
                Thread.sleep(300);
                HttpResponse<JsonNode> jsonResponse = Unirest. //
                        get(COUNTRY_INFORMATION_SERVICE_URL + "&offset=" + offset). //
                        header("accept", "application/json"). //
                        header("Authorization", "Bearer " + System.getenv("RESTCOUNTRIES_API_KEY")). //
                        asJson();

                String body = jsonResponse.getBody().toString();
                TypeReference<Root> typeRef = new TypeReference<Root>() {
                };
                Root root = objectMapper.readValue(body, typeRef);
                countryDescriptions.addAll(root.data.objects);
            }

        } catch (Exception e) {
            throw new RestCountriesAPIException("Could not read country information from " + COUNTRY_INFORMATION_SERVICE_URL, e);
        }

        return countryDescriptions;
    }

}

File: TypeScript/app/d/rest-countries-api.ts
import { Country } from "./country";

import fetch from 'node-fetch';

interface RestCountry {
    codes: {
        alpha_2: string;
    };
    region: string;
    coordinates: {
        lat: number;
        lng: number;
    };
}

export class RestCountriesAPI {
    private readonly HOME_BASE: Country = { name: 'AT' };
    private readonly COUNTRY_INFORMATION_SERVICE_URL: string = 'https://api.restcountries.com/countries/v5?pretty=1&limit=100';

    static getCurrent(): RestCountriesAPI {
        return new RestCountriesAPI();
    }

    private constructor() {

    }

    public isInCommonMarket(country: Country) {
        const countryDescription = this.getCountryDescriptionViaRestCall(country);
        return countryDescription.then((c) => c?.region === 'Europe');
    }

    public isInAmericas(country: Country) {
        const countryDescription = this.getCountryDescriptionViaRestCall(country);
        return countryDescription.then((c) => c?.region === 'Americas');
    }

    public distanceTo(country: Country): Promise<number> {
        return Promise.all([
            this.getCountryDescriptionViaRestCall(this.HOME_BASE),
            this.getCountryDescriptionViaRestCall(country)
        ]).then(bothCountries => {
            if (!bothCountries || !bothCountries[0] || !bothCountries[1]) {
                return 0;
            }
            return this.distBetween(bothCountries[0].coordinates.lat, bothCountries[0].coordinates.lng,
                bothCountries[1].coordinates.lat, bothCountries[1].coordinates.lng);
        });
    }

    private distBetween(fromLatitude: number, fromLongitude: number, toLatitude: number, toLongitude: number): number {
        const earthRadius = 6371000; // meters
        const fLat: number = this.toRadians(fromLatitude);
        const toLat: number = this.toRadians(toLatitude);
        const diffLat: number = this.toRadians(toLatitude - fromLatitude);
        const diffLng: number = this.toRadians(toLongitude - fromLongitude);
        const a = Math.sin(diffLat / 2) * Math.sin(diffLat / 2)
            + Math.cos(fLat) * Math.cos(toLat) * Math.sin(diffLng / 2) * Math.sin(diffLng / 2);
        const c: number = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));
        const dist: number = (earthRadius * c);
        return Math.floor(dist);
    }

    private toRadians(value: number): number {
        return value * Math.PI / 180;
    }

    private async getCountryDescriptionViaRestCall(country: Country): Promise<RestCountry | null> {
        const countryDescriptions: RestCountry[] = await this.slowHttpCall();

        for (let i = 0; i < countryDescriptions.length; i++) {
            if (countryDescriptions[i].codes.alpha_2 === country.name) {
                return countryDescriptions[i];
            }
        }

        return null;
    }

    private async slowHttpCall(): Promise<RestCountry[]> {
        const sleep = (ms: number) => new Promise((resolve) => setTimeout(resolve, ms));

        let countryDescriptions: RestCountry[] = [];

        for (let offset = 0; offset < 300; offset += 100) {
            await sleep(300); // Sleep for 1 second
            const headers = {
                'accept': 'application/json',
                'Authorization': 'Bearer ' + process.env.RESTCOUNTRIES_API_KEY
            };
            const response = await fetch(this.COUNTRY_INFORMATION_SERVICE_URL + '&offset=' + offset,
                {
                    method: 'GET',
                    headers
                }
            );
            const jsonBody = JSON.parse(await response.text());
            countryDescriptions = countryDescriptions.concat(jsonBody.data.objects);
        }

        return countryDescriptions;
    }

}
